apps/arweave/src/ar_mining_worker.erl

```diff
@@ -245,362 +284,412 @@ terminate(_Reason, _State) ->
-			%% 2. chunk2 that was just read and will shortly be used to compute h2
-			update_sub_chunk_cache_size(-2, SessionKey, State2);
-		{{chunk1, H1}, State2} ->
-			ar_mining_hash:compute_h2(self(), Candidate2#mining_candidate{ h1 = H1 }),
-			%% Decrement 1 for chunk2:
-			%% we're computing h2 for a peer so chunk1 was not previously read or cached 
-			%% on this node
-			update_sub_chunk_cache_size(-1, SessionKey, State2);
-		{do_not_cache, State2} ->
-			%% Decrement 1 for chunk2
-			%% do_not_cache indicates chunk1 was not and will not be read or cached
-			update_sub_chunk_cache_size(-1, SessionKey, State2);
-		{cached, State2} ->
-			case Candidate2#mining_candidate.cm_lead_peer of
-				not_set ->
-					ok;
-				_ ->
-					?LOG_ERROR([{event, cm_chunk2_cached_before_chunk1},
-						{worker, State#state.name},
-						{partition_number, Candidate2#mining_candidate.partition_number},
-						{partition_number2, Candidate2#mining_candidate.partition_number2},
-						{cm_peer, ar_util:format_peer(Candidate2#mining_candidate.cm_lead_peer)},
-						{cache_ref, Candidate2#mining_candidate.cache_ref},
-						{nonce, Candidate2#mining_candidate.nonce},
-						{session, ar_nonce_limiter:encode_session_key(SessionKey)}])
-			end,
-			State2
-	end.
-
-priority(computed_h2, StepNumber) ->
-	{1, -StepNumber};
-priority(computed_h1, StepNumber) ->
-	{2, -StepNumber};
-priority(compute_h2_for_peer, StepNumber) ->
-	{2, -StepNumber};
-priority(chunk2, StepNumber) ->
-	{3, -StepNumber};
-priority(chunk1, StepNumber) ->
-	{4, -StepNumber};
-priority(computed_h0, StepNumber) ->
-	{5, -StepNumber};
-priority(compute_h0, StepNumber) ->
-	{6, -StepNumber}.
-
-handle_task({chunk1, Candidate, [RangeStart, ChunkOffsets]}, State) ->
-	State2 = process_chunks(chunk1, Candidate, RangeStart, ChunkOffsets, State),
-	{noreply, State2};
-
-handle_task({chunk2, Candidate, [RangeStart, ChunkOffsets]}, State) ->
-	State2 = process_chunks(chunk2, Candidate, RangeStart, ChunkOffsets, State),
-	{noreply, State2};
+-spec handle_task(
+	Task :: {
+		EventType :: compute_h0 | computed_h0 | chunk1 | chunk2 | computed_h1 | computed_h2 | compute_h2_for_peer,
+		Candidate :: #mining_candidate{},
+		ExtraArgs :: term()
+	},
+	State :: #state{}
+) -> State :: #state{}.
 
+%% @doc Handle the `compute_h0` task.
+%% Indicates that the VDF step has been computed.
 handle_task({compute_h0, Candidate, _ExtraArgs}, State) ->
-	#state{ latest_vdf_step_number = LatestVDFStepNumber,
-			vdf_queue_limit = VDFQueueLimit } = State,
-	#mining_candidate{ session_key = SessionKey, step_number = StepNumber } = Candidate,
-	State2 = report_hashes(State),
-	State4 = case try_to_reserve_cache_space(SessionKey, State2) of
-		{true, State3} ->
+	#state{
+		latest_vdf_step_number = LatestVDFStepNumber,
+		vdf_queue_limit = VDFQueueLimit
+	} = State,
+	#mining_candidate{ step_number = StepNumber } = Candidate,
+	State1 = report_and_reset_hashes(State),
+	% Check if we need to compute h0 early
```

Comment:
```suggestion
	% Only mine recent VDF Steps
```

Previously the intent of this line wasn't to compute h0 early, but decide whther or not to compute this VDF step at all. If StepNumber is too old, then we just drop it so that we don't waste CPU and DiskIO hashing an old VDF step that's likely to generate solutions with a higher probability of being rejected by peers.